File: app/src/main/java/com/lytefast/flexinput/sampleapp/MessageAdapter.java
package com.lytefast.flexinput.sampleapp;

import android.net.Uri;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.interfaces.DraweeController;
import com.facebook.drawee.view.SimpleDraweeView;

import java.util.ArrayList;
import java.util.List;

import com.lytefast.flexinput.model.Attachment;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;


/**
 * Simple string message display adapter.
 *
 * @author Sam Shih
 */

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.ViewHolder> {
  private final List<Data> msgList = new ArrayList<>();

  @Override
  public ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
    View view = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.message_row, parent, false);
    return new ViewHolder(view);
  }

  @Override
  public void onBindViewHolder(final ViewHolder holder, final int position) {
    holder.bind(msgList.get(position), position);
  }

  @Override
  public int getItemCount() {
    return msgList.size();
  }

  public void addMessage(Data msg) {
    msgList.add(msg);
    notifyItemInserted(msgList.size() - 1);
  }

  public static class Data {
    final Editable editable;
    @Nullable
    final Attachment attachment;

    public Data(Editable editable, Attachment attachment) {
      this.editable = editable;
      this.attachment = attachment;
    }
  }

  static class ViewHolder extends RecyclerView.ViewHolder {
    private final TextView indexTv;
    private final TextView messageTv;
    private final TextView attachmentTv;
    private final SimpleDraweeView imageView;

    ViewHolder(final View itemView) {
      super(itemView);

      indexTv = itemView.findViewById(R.id.index_tv);
      messageTv = itemView.findViewById(R.id.message_tv);
      attachmentTv = itemView.findViewById(R.id.attachment_tv);
      imageView = itemView.findViewById(R.id.attachment_iv);
    }

    void bind(Data data, int index) {
      indexTv.setText(String.valueOf(index));
      messageTv.setText(data.editable);
      if (data.attachment != null) {
        imageView.setVisibility(View.VISIBLE);
        Uri uri = data.attachment.getUri();

        DraweeController controller = Fresco.newDraweeControllerBuilder()
            .setUri(uri)
            .setAutoPlayAnimations(true)
            .build();
        imageView.setController(controller);

        attachmentTv.setVisibility(View.VISIBLE);
        attachmentTv.setText(data.attachment.getDisplayName());
      } else {
        imageView.setVisibility(View.GONE);
        attachmentTv.setVisibility(View.GONE);
      }
    }
  }
}

File: app/src/main/java/com/lytefast/flexinput/sampleapp/MessageAdapter.kt
package com.lytefast.flexinput.sampleapp

import android.text.Editable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.facebook.drawee.backends.pipeline.Fresco
import com.facebook.drawee.interfaces.DraweeController
import com.facebook.drawee.view.SimpleDraweeView
import com.lytefast.flexinput.model.Attachment
import java.util.*

/**
 * Simple string message display adapter.
 *
 * @author Sam Shih
 */
class MessageAdapter : RecyclerView.Adapter<MessageAdapter.ViewHolder>() {
  private val msgList: MutableList<Data> = ArrayList()
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    val view = LayoutInflater.from(parent.context)
        .inflate(R.layout.message_row, parent, false)
    return ViewHolder(view)
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    holder.bind(msgList[position], position)
  }

  override fun getItemCount(): Int {
    return msgList.size
  }

  fun addMessage(msg: Data) {
    msgList.add(msg)
    notifyItemInserted(msgList.size - 1)
  }

  class Data(val editable: Editable, val attachment: Attachment<*>?)

  class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

    private val indexTv: TextView = itemView.findViewById(R.id.index_tv)
    private val messageTv: TextView = itemView.findViewById(R.id.message_tv)
    private val attachmentTv: TextView = itemView.findViewById(R.id.attachment_tv)
    private val imageView: SimpleDraweeView = itemView.findViewById(R.id.attachment_iv)

    fun bind(data: Data, index: Int) {
      indexTv.text = index.toString()
      messageTv.text = data.editable
      if (data.attachment != null) {
        imageView.visibility = View.VISIBLE
        val uri = data.attachment.uri
        val controller: DraweeController = Fresco.newDraweeControllerBuilder()
            .setUri(uri)
            .setAutoPlayAnimations(true)
            .build()
        imageView.controller = controller
        attachmentTv.visibility = View.VISIBLE
        attachmentTv.text = data.attachment.displayName
      } else {
        imageView.visibility = View.GONE
        attachmentTv.visibility = View.GONE
      }
    }
  }
}